src/kademlia/routing_table.cpp

```diff
@@ -682,13 +701,37 @@ routing_table::add_node_status_t routing_table::add_node_impl(node_entry e)
 		else
 		{
 			TORRENT_ASSERT(existing->id != e.id);
-			// this is the same IP and port, but with
-			// a new node ID. remove the old entry and
-			// replace it with this new ID
+			// This is the same IP and port, but with a new node ID.
+			// This may indicate a malicious node so remove the entry.
+#ifndef TORRENT_DISABLE_LOGGING
+			char hex_id_new[41];
+			char hex_id_old[41];
+			aux::to_hex(reinterpret_cast<char const*>(&e.id[0]), 20, hex_id_new);
+			aux::to_hex(reinterpret_cast<char const*>(&existing->id[0]), 20, hex_id_old);
+			m_log->log(dht_logger::routing_table, "evicting node (changed ID): old: %s new: %s %s"
+				, hex_id_old, hex_id_new, print_address(e.addr()).c_str());
+#endif
+
 			remove_node(existing, existing_bucket);
+			fill_from_replacements(existing_bucket);
+
+			// when we detect possible malicious activity in a bucket,
+			// schedule the other nodes in the bucket to be pinged soon
+			// to clean out any other malicious nodes
+			auto now = aux::time_now();
+			for (auto& node : existing_bucket->live_nodes)
+			{
+				if (now - node.last_queried > minutes(5))
+					node.last_queried = min_time();
+			}
+
+			return failed_to_add;
```

Comment:
is it possible to trick libtorrent (with this patch) to remove a bunch of nodes by sending spoofed packets, with their source (IP,port)?